File: Leetcode/Hard/1770. Maximum Score from Performing Multiplication Operations.cpp
class Solution {
public:
    vector<vector<int>> dp;
    int maximumScore(vector<int>& nums, vector<int>& multipliers) {
        dp = vector<vector<int>>(nums.size()+1, vector<int>(multipliers.size()+1,1e9));
        return dfs(nums,multipliers, 0, 0);
    }

    int dfs(vector<int>& nums, vector<int>& multipliers, int l, int opt){
        if(opt >= multipliers.size()){
            return 0;
        }
        if(dp[l][opt] != 1e9){
            return dp[l][opt];
        }
        int left = multipliers[opt] * nums[l] + dfs(nums,multipliers,l+1,opt+1);
        int r = nums.size() - 1 - (opt - l);
        int right = multipliers[opt] * nums[r] + dfs(nums,multipliers, l, opt+1);
        dp[l][opt] = max(left,right);
        return dp[l][opt];
    }
};

File: Leetcode/Hard/1770. Maximum Score from Performing Multiplication Operations/1770. Maximum Score from Performing Multiplication Operations.py
class Solution:
    def maximumScore(self, nums: List[int], multipliers: List[int]) -> int:
        n1, n2 = len(nums), len(multipliers)
        memo = {}
        
        def dfs(l, idx):
            if idx >= n2:
                return 0

            if (l, idx) in memo:
                return memo[(l, idx)]
                
            r = n1 - 1 - (idx - l)
            
            left = nums[l] * multipliers[idx] + dfs(l + 1, idx + 1)
            right = nums[r] * multipliers[idx] + dfs(l, idx + 1)

            memo[l, idx] = max(left, right)
            return memo[(l, idx)]

        return dfs(0, 0)